cns/kubecontroller/nodenetworkconfig/reconciler.go

```diff
@@ -74,6 +75,15 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
 	ipAssignments := 0
 
+	// during node upgrades, an nnc may be updated with new ncs. at any given time, only the ncs
+	// that exist in the nnc are valid. any others that may have been previously created and no
+	// longer exist in the nnc should be considered stale.
+	validNCIDs := make([]string, len(nnc.Status.NetworkContainers))
```

Comment:
This should be done only during CNS restart at the time it is reconciling with CNI/NNC state, not here. (or may be here too, still thinking).

Reply:
The issue arises *because* CNS doesn't restart, no?

Reply:
Yes we should handle cleaning up this stale NC during initial reoncile flow and before actually starting http listener.

Reply:
wait, if it is node reimage then i would expect CNS to start no?

Reply:
regardless of the CNS restart, this should be enough to take care of the NC updates in NNC and remove stale NCs, no? @rbtr ?
Do we not execute reconcile upon CNS restart too?

Reply:
IMO we don't need to do this at startup, we have to do it for each status update anyway so when reconciler starts, it will get an event for this status and do the cleanup